### PR TITLE
Fix 1246: configure default order on ApiResource annotation

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1246
| License       | MIT
| Doc PR        | api-platform/doc#235

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/Extension/OrderExtension.php [hunk 1]
```diff
@@ -14,20 +14,24 @@
...
  * @author Samuel ROZE <samuel.roze@gmail.com>
+ * @author Vincent Chalamon <vincentchalamon@gmail.com>
  */
 final class OrderExtension implements QueryCollectionExtensionInterface
 {
     private $order;
+    private $resourceMetadataFactory;
 
-    public function __construct(string $order = null)
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, string $order = null)
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/OrderExtension.php]: This a BC break, can you add it as an optional last parameter?

**Comment by Simperfit** [comment]
Could you please rebase instead of merging ?

**Comment by dunglas** [comment]
Thank you @vincentchalamon

File: tests/Fixtures/TestBundle/Entity/Foo.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
+ * Foo.
+ *
+ * @author Vincent Chalamon <vincentchalamon@gmail.com>
+ *
+ * @ApiResource(attributes={
+ *     "order"={"name", "DESC"}
```
Comment (dunglas) [inline tests/Fixtures/TestBundle/Entity/Foo.php]: I probably have been too fast to merge. Wouldn't be better to allow complex ordering? I suggest the following:

```php
@ApiResource(attributes={"name": "DESC", "firstname": "ASC"})
```

(Order by name then by firstname). WDYT @vincentchalamon? Can you do a follow up PR?

Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Foo.php]: I though about adding the multiple default ordering, but I wasn't sure it was very useful.
I would prefer a syntax like that:
```php
@ApiResource(attributes={
    "order"={"name": "DESC", "firstname": "ASC"}
})
```
WDYT @dunglas? If so, I could open another PR to fix it.

Comment (dunglas) [inline tests/Fixtures/TestBundle/Entity/Foo.php]: IMO it's useful (ordering by name then firstname is very common for instance). You're right for the syntax, my example is faulty (missing `order` key). A PR would be great!

Comment (vincentchalamon) [inline tests/Fixtures/TestBundle/Entity/Foo.php]: https://github.com/api-platform/core/pull/1323

**Comment by ahmed-bhs** [comment]
Guys thank you for this feature, does this support yaml configuration, I tried to add order to my resource, seems nothing work ! Any idea? @vincentchalamon 
```
resources:
    App\Entity\Repos\EventFamily:
        collectionOperations:
            get:
                order: ["name", "DESC"]
                attributes:
                    order: ["name", "DESC"]
```

**Comment by vincentchalamon** [comment]
@ahmed-bhs Did you try the following configuration?:
```yaml
resources:
    App\Entity\Repos\EventFamily:
        order: ["name", "DESC"]
        collectionOperations:
            get: ~
```

I didn't try this configuration in YAML since many years (this PR is older than 3 years).

**Pull request merged**